visions/webassembly.md

```diff
@@ -0,0 +1,129 @@
+examples of its possible adoption in the Swift toolchain itself. To quote
+[a GSoC 2024 idea](https://www.swift.org/gsoc2024/#building-swift-macros-with-webassembly):
+
+> WebAssembly could provide a way to build Swift macros into binaries that can be distributed and run anywhere,
+> eliminating the need to rebuild them continually.
+
+This can be applicable not only to Swift macros, but also for the evaluation of SwiftPM manifests and plugins.
+
+The WebAssembly instruction set has useful properties from a security perspective, as it has
+no interrupts or peripheral access instructions. Access to the underlying system is always done by calling
+explicitly imported functions, implementations for which are provided by an imported WebAssembly module or a WebAssembly
+runtime itself. The runtime has full control over interactions of the virtual machine with the outside world.
+
+WebAssembly code and data live in completely separate address spaces, with all executable code in a given module loaded
+and validated by the runtime upfront. Combined with the lack of "jump to address" and a limited set of control flow
+instructions that require explicit labels in the same function body, this makes a certain class of attacks impossible to
+execute in a correctly implemented spec-compliant WebAssembly runtime.
+
+In the context of Swift developer tools, arbitrary code execution during build time can be virtualized with Wasm.
+While Swift macros, SwiftPM manifests, and plugins are sandboxed on Darwin platforms, with Wasm we can provide stronger
+security guarantees on other platforms that have a compatible Wasm runtime available.
+
+The WebAssembly instruction set is designed with performance in mind. A WebAssembly module can be JIT-compiled or
+compiled on a client machine to an optimized native binary ahead of time. With recently accepted proposals to the Wasm
+specification it now supports features such as SIMD, atomics, multi-threading, and more. A WebAssembly runtime can
+generate a restricted subset of native binary code that implements these features with little performance overhead.
+
+This means that adoption of Wasm in developer tools does not imply unavoidable performance overhead. In fact, with
+security guarantees that virtualization brings there's no longer a need to spawn a separate process for each
+Swift compiler and SwiftPM plugin/manifest invocation. Virtualized Wasm binaries can run in the host process of a
+Wasm runtime, removing the overhead of new process setup and IPC infrastructure.
+
+### WebAssembly System Interface and the Component Model
+
+The WebAssembly instruction set on its own doesn't "support" file I/O or networking, in the same way that ARM64 or x86_64
+don't "support" those directly either. Actual implementation of I/O for a hardware CPU is provided by the operating
+system, and for a Wasm module it's provided by a runtime that executes it.
```

Comment:
We should remove the quotes around "support". Also, the example of ARM64 or x86_64 not supporting file I/O or networking directly and requiring an operating system isn't entirely true; I/O for hardware CPUs is typically done by mapping devices into their physical memory map. (x86 is an especially bad choice of example also because of features like IPMI and AMT.)

```suggestion
The WebAssembly virtual machine has no in-built support for I/O; instead, a Wasm module's access to I/O is dependent entirely upon the runtime that executes it.
```